### PR TITLE
dispatch blob messages as arraybuffers

### PR DESCRIPTION
**Pull request by trevj**
Dispatches messages received as Blobs as ArrayBuffers, to work around the following issue in Firefox add-ons:
https://bugzilla.mozilla.org/show_bug.cgi?id=1122682

Depends on:
 * https://github.com/willscott/webrtc-adapter/pull/3
 * https://github.com/freedomjs/freedom-for-firefox/pull/45

### CODE REVIEW

**Comment by willscott** [comment]
Only question is whether we can fix this problem purely in webrtc-adapter as in my comment there (https://github.com/willscott/webrtc-adapter/pull/3#issuecomment-70729371). I'm not sure we can, in which case this seems fine.

**Comment by trevj** [comment]
I've added a test to send binary data across the datachannel in integration tests, similar to what we're already doing for string messages.

I couldn't think of a way to achieve this solely via the adapter :-/

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 1]
```diff
@@ -54,8 +59,28 @@ module.exports = function (pc, dc, setup) {
     }, function (err) {
       console.error('RTC failed: ',err);
     });
-  });
+  };
   
+  it("Sends string messages across data channels", function (done) {
+    sendMessageToPeer('message from alice', function(result) {
+      expect(result.text).toEqual('message from alice');
+      done();
```
Comment (willscott) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: probably should call done() after closing the channels, i think that would just be moving the callback(msg) after the bob.close() above

Comment (trevj) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: Done.

File: providers/core/core.rtcdatachannel.js [hunk 2]
```diff
@@ -124,9 +125,16 @@ RTCDataChannelAdapter.prototype.onclose = function (event) {
 RTCDataChannelAdapter.prototype.onmessage = function (event) {
   if (typeof event.data === 'string') {
     this.dispatchEvent('onmessage', {text: event.data});
-  } else {
+  } else if (event.data instanceof ArrayBuffer) {
     this.dispatchEvent('onmessage', {buffer: event.data});
+  } else if (event.data instanceof Blob) {
```
Comment (bemasc) [inline providers/core/core.rtcdatachannel.js]: ` && this.channel.binaryType == 'arraybuffer'`

Comment (trevj) [inline providers/core/core.rtcdatachannel.js]: Very good point although if Freedom can't support `Blob` then perhaps it should always convert to an `ArrayBuffer`?

**Comment by trevj** [comment]
So, two updates:
 1. I merged your latest fixes into this branch.
 1. I moved the Blob conversion function(s) I was proposing for the adapter into here.

Still hashing out exactly what's happening regarding Blob + FileReader + JSM/non-content scripts with the Firefox guys but now the `Sends binary messages across data channels passed` test passes in Firefox (FF 35, Linux) *and* Simple SOCKS runs!

**Comment by willscott** [comment]
can you add the updated test as part of this?

**Comment by trevj** [comment]
Updated test? There's a new test which sends binary data. It doesn't pass
in FF without the changes to core.rtcdatachannel.

**Comment by willscott** [comment]
whoops, was looking at an outdated diff

**Comment by willscott** [comment]
Okay, so here's my question:

Forcing binarytype to `arraybuffer`, as in
https://github.com/freedomjs/freedom/commit/e77ede7e2362f2269333e8a1a567d5a7177975cc
Makes this test pass for me in firefox 34/mac.  I'll test on linux when I get home this evening.
If that's all we need to do, that seems simpler than adding the manual blob->arraybuffer conversion.

Does solving this by making the RTC client explicitly request arraybuffers work?

**Comment by trevj** [comment]
Thanks for the patch!

No, *the test does not pass* on my system (Linux/Firefox 35.0.1).

I get the following output:
```
>> integration: core.rtcpeerconnection Sends binary messages across data channels failed
>> Expected 0 to equal 5.
```

Fascinating! I guess this helps explain our confusion...I had indeed tried exactly that patch before without any luck. Hooray for stable tests :-)

**Comment by trevj** [comment]
OK, so, I'll add a call to set `binaryType = 'arraybuffer'` and we should keep the blob conversion functions for those platforms which need it. Sound good?

**Comment by trevj** [comment]
OK, this just gets more fun.

I made `core.rtcdatachannel` always set the `binaryType` of new data channels to 'arraybuffer':
https://github.com/freedomjs/freedom/commit/9a65fed91abd85535641531672930ea0afb73507

Result? The integration tests break on those versions of Firefox with the `ArrayBuffer` weirdness in add-ons! Basically, I get the weird "opaque ArrayBuffer" thing I reported here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1122687

What if I *also* try importing `ArrayBuffer` from `hiddenWindow`? Then I get *crazy* errors trying to work with ArrayBuffer elsewhere in my app, e.g.:
```
JavaScript error: resource://jid1-jbosl7j4ljmkzw-at-jetpack/simple-socks/data/lib/socks-common/socks-headers.js, line 90: Error: unsupported SOCKS version: 0
```

So, I think for now the best we can do is always convert blobs when we find them. PTAL :-)

**Comment by willscott** [comment]
This seems different though from asking the client to call setBinaryType()
which seems to be working in the current tests.
Can we just do that?

--Will

**Comment by trevj** [comment]
Huh? No, we can't. Not if you want stuff to work in Firefox add-ons, anyway.

**Comment by trevj** [comment]
Well, you certainly can ask clients to call `setBinaryType` -- uProxy won't be calling it though.

Is that what you mean?

**Comment by willscott** [comment]
Right, that was what I meant. Does that have the same issues of confusing
some platforms as setting binary type in freedom synchronous with
datachannel creation?

**Comment by trevj** [comment]
I think it will have the same issues, yes.

**Comment by willscott** [comment]
Trying this on firefox 35 / linux, and can replicate the weirdness.

Do we have any idea what the actual object is that comes out of
`dataChannel.onMessage`? It doesn't have `size` or `type` properties that a
blob claims to - which leaves me generally confused about what this object
actually is (this is in the core provider before it's gone through any
freedom conversions)

**Comment by bemasc** [comment]
It's an event

**Comment by willscott** [comment]
Was talking about the event.data property that is provided to that event,
and acts like neither a blob nor arraybuffer.

--will

**Comment by trevj** [comment]
I don't know what it is but we seem to be getting it because of the varying scopes in play:
https://bugzilla.mozilla.org/show_bug.cgi?id=1122687

My suggestion is to merge this pull request and continue to investigate the `Blob`/`ArrayBuffer` weirdness, raising the priority if performance proves an issue.

**Comment by trevj** [comment]
As discussed, perhaps we could be a little smarter about when to do conversion.

However, I do see the `size` and `type` properties on the "blob" object. Specifically, I see:

```
slice
size
type
```

So, I guess it's not `Blob` that's the problem: it's `ArrayBuffer`.

This means we can only test for the presence of the bug when we try to convert a `Blob` to an `ArrayBuffer`. Fine. I've updated the code to only convert to `ArrayBuffer` if a `Blob` is received and, furthermore, only convert via a binary string if `FileReader.readAsArrayBuffer` returns a non-`ArrayBuffer`.

File: providers/core/core.rtcdatachannel.js (hunk 2 above)
Comment (bemasc) [inline providers/core/core.rtcdatachannel.js]: @willscott, what's the status of Blob and Freedom?

**Comment by trevj** [comment]
Ping!

Could really do with getting this in.

**Comment by willscott** [comment]
Okay - played with this again for a few minutes this evening and made progress:

If you ask for binary type `arraybuffer` in firefox, you sometimes get an opaque object, as we've noticed.

i tried to see if I could ask for a `Uint8Array(event.data)`, and firefox throws a super helpful message:
```
JavaScript error: Accessing TypedArray data over Xrays is slow, and forbidden in order to encourage performant code. To copy TypedArrays across origin boundaries, consider using Components.utils.cloneInto().
```

The following change to `RTCDataChannelAdapter.prototype.onmessage`, plus requesting `setBinaryType('arraybuffer')` seems to work:
```javascript
if(this.channel.binaryType === 'arraybuffer' && typeof Components !== 'undefined' && !(event.data instanceof ArrayBuffer)) {
  var view = {};
  var realData = Components.utils.cloneInto(event.data, view);
  this.dispatchEvent('onmessage', {buffer: realData});
} else {
  this.dispatchEvent('onmessage', {buffer: event.data});
}
```

This feels a bit cleaner than introducing the blob->arraybuffer conversion and requesting blobs on firefox and then still passing it / expecting arraybuffers.

**Comment by willscott** [comment]
For reference: https://developer.mozilla.org/en-US/docs/Components.utils.cloneInto

**Comment by trevj** [comment]
Did you mean something like this?:
https://github.com/freedomjs/freedom/compare/trevj-blobs-clone-suggestion?expand=1

This doesn't pass integration tests on my system (Ubuntu 14, Firefox 35) or make Simple SOCKS work.

**Comment by willscott** [comment]
@trevj  Here's what I meant: https://github.com/freedomjs/freedom/pull/223  Your branch was only looking for things that were blobs, so wouldn't have been running on the opaque objects if my read is correct.

**Comment by willscott** [comment]
Went with https://github.com/freedomjs/freedom/pull/223 instead.

**Pull request closed without merging**